casts/highlights/src/UIHighlights.tsx

```diff
@@ -109,53 +74,69 @@ export function UIHighlights({
     pagingEnabled = false,
     controlsHidden = false,
 }: UIHighlightsProps) {
+    const scrollRef = useAnimatedRef<Animated.ScrollView>();
+
+    const scrollViewWidthShared = useSharedValue(0);
+    const scrollViewContentWidthShared = useSharedValue(0);
+
     const currentGravityPosition = useSharedValue(0);
     const currentProgress = useSharedValue(0);
 
+    const isRtlShared = useSharedValue(I18nManager.getConstants().isRTL);
+
     const {
         onItemLayout,
         calculateCurrentPosition,
         calculateClosestX,
-        calculateClosestLeftX,
-        calculateClosestRightX,
-        sharedContext,
-    } = usePositions(React.Children.count(children));
+        calculateClosestPreviousX,
+        calculateClosestNextX,
+    } = usePositions(
+        React.Children.count(children),
+        scrollViewContentWidthShared,
+        scrollViewWidthShared,
+        isRtlShared,
+    );
 
-    const scrollRef = useAnimatedRef<Animated.ScrollView>();
+    const onScrollViewLayout = React.useCallback(
+        (event: LayoutChangeEvent) => {
+            const {
+                nativeEvent: {
+                    layout: { width },
+                },
+            } = event;
 
-    const scrollViewWidthRef = React.useRef<number>(0);
-    const scrollViewContentWidthRef = React.useRef<number>(0);
-    const onScrollViewLayout = React.useCallback((event: LayoutChangeEvent) => {
-        const {
-            nativeEvent: {
-                layout: { width },
-            },
-        } = event;
-        scrollViewWidthRef.current = width;
-    }, []);
-
-    const onScrollViewSizeChange = React.useCallback((w: number) => {
-        scrollViewContentWidthRef.current = w;
-    }, []);
+            scrollViewWidthShared.value = width;
+        },
+        [scrollViewWidthShared],
+    );
+
+    const onScrollViewSizeChange = React.useCallback(
+        (w: number) => {
+            scrollViewContentWidthShared.value = w;
+        },
+        [scrollViewContentWidthShared],
+    );
 
     type ScrollContext = { adjustOnMomentum: number };
     const scrollHandler = useAnimatedScrollHandler<ScrollContext>({
         onScroll(event) {
             const {
-                contentOffset: { x },
+                contentOffset: { x: rawX },
             } = event;
 
-            if (x == null || isNaN(x)) {
+            if (rawX == null || isNaN(rawX)) {
                 return;
             }
 
-            // Check if we've reached the end
-            if (x >= scrollViewContentWidthRef.current - scrollViewWidthRef.current) {
-                // Set the current gravity position as for the last item
-                currentGravityPosition.value = sharedContext.value.positions.length - 1;
-                currentProgress.value = 0;
-                return;
-            }
```

Comment:
AFAIR this was very important fix to show the dot when the last card could not be scrolled to the left border of the visible view.

Don't we need it anymore?

Reply:
You are right. I missed this. I added this logic to [usePosition.tsx](https://github.com/tonlabs/UIKit/pull/1695/files#diff-9f4515115482259d6c48fe751437cd954d109a84db48dd2b576519e889abbdbaR113-R123)